agent/agent.go

```diff
@@ -639,13 +641,77 @@ func (a *agent) runScript(ctx context.Context, lifecycle, script string) error {
 	}
 
 	a.logger.Info(ctx, "running script", slog.F("lifecycle", lifecycle), slog.F("script", script))
-	writer, err := a.filesystem.OpenFile(filepath.Join(a.logDir, fmt.Sprintf("coder-%s-script.log", lifecycle)), os.O_CREATE|os.O_RDWR, 0o600)
+	fileWriter, err := a.filesystem.OpenFile(filepath.Join(a.logDir, fmt.Sprintf("coder-%s-script.log", lifecycle)), os.O_CREATE|os.O_RDWR, 0o600)
 	if err != nil {
 		return xerrors.Errorf("open %s script log file: %w", lifecycle, err)
 	}
 	defer func() {
-		_ = writer.Close()
+		_ = fileWriter.Close()
 	}()
+
+	startupLogsReader, startupLogsWriter := io.Pipe()
+	defer func() {
+		_ = startupLogsReader.Close()
+		_ = startupLogsWriter.Close()
+	}()
+	writer := io.MultiWriter(startupLogsWriter, fileWriter)
+
+	queuedLogs := make([]agentsdk.StartupLog, 0)
+	var flushLogsTimer *time.Timer
+	var logMutex sync.Mutex
+	var logsSending bool
+	sendLogs := func() {
+		logMutex.Lock()
+		if logsSending {
+			logMutex.Unlock()
+			return
+		}
+		logsSending = true
+		logsToSend := queuedLogs
+		queuedLogs = make([]agentsdk.StartupLog, 0)
+		logMutex.Unlock()
+		for r := retry.New(time.Second, 5*time.Second); r.Wait(ctx); {
+			err := a.client.PatchStartupLogs(ctx, agentsdk.PatchStartupLogs{
+				Logs: logsToSend,
+			})
```

Comment:
it's possible our error here is due to exceeding our log limit in which case retrying doesn't really help us. Maybe we should check for the error and then `io.Discard` the rest of the output from the pipe